orchagent/cbf/cbfnhgorch.cpp

```diff
@@ -0,0 +1,799 @@
+                            status);
+            return false;
+        }
+
+        /* Update the selection map and update the previous and new map ref count. */
+        gNhgMapOrch->decRefCount(m_selection_map);
+        m_selection_map = selection_map;
+        gNhgMapOrch->incRefCount(m_selection_map);
+    }
+
+    return true;
+}
+
+/*
+ * Purpose: Sync the given CBF group members.
+ *
+ * Params:  IN members - The members to sync.
+ *
+ * Returns: true, if the operation was successful,
+ *          false, otherwise.
+ */
+bool CbfNhg::syncMembers(const set<string> &members)
+{
+    SWSS_LOG_ENTER();
+
+    /* The group should be synced at this point. */
+    if (!isSynced())
+    {
+        SWSS_LOG_ERROR("Trying to sync members of CBF next hop group %s which is not synced",
+                       m_key.c_str());
+        throw logic_error("Syncing members of unsynced CBF next hop group");
+    }
+
+    /*
+     * Sync all the given members.  If a NHG does not exist, is not yet synced
+     * or is temporary, stop immediately.
+     */
+    ObjectBulker<sai_next_hop_group_api_t> bulker(sai_next_hop_group_api,
+                                                  gSwitchId,
+                                                  gMaxBulkSize);
+    unordered_map<string, sai_object_id_t> nhgm_ids;
+
+    for (const auto &key : members)
+    {
+        auto &nhgm = m_members.at(key);
+
+        /* A next hop group member can't be already synced if it was passed into the syncMembers() method. */
+        if (nhgm.isSynced())
+        {
+            SWSS_LOG_ERROR("Trying to sync already synced CBF NHG member %s in group %s",
+                           nhgm.to_string().c_str(), to_string().c_str());
+            throw std::logic_error("Syncing already synced NHG member");
+        }
+
+        /*
+         * Check if the group exists in NhgOrch.
+         */
+        if (!gNhgOrch->hasNhg(key))
+        {
+            SWSS_LOG_ERROR("Next hop group %s in CBF next hop group %s does "
+                            "not exist", key.c_str(), m_key.c_str());
+            return false;
+        }
+
+        const auto &nhg = gNhgOrch->getNhg(key);
+
+        /*
+         * Check if the group is synced.
+         */
+        if (!nhg.isSynced())
+        {
+            SWSS_LOG_ERROR("Next hop group %s in CBF next hop group %s is not"
+                            " synced",
+                            key.c_str(), m_key.c_str());
+            return false;
+        }
+
+        /* Create the SAI attributes for syncing the NHG as a member. */
+        auto attrs = createNhgmAttrs(nhgm);
+
+        bulker.create_entry(&nhgm_ids[key],
+                            (uint32_t)attrs.size(),
+                            attrs.data());
+    }
+
+    /* Flush the bulker to perform the sync. */
+    bulker.flush();
+
+    /* Iterate over the synced members and set their SAI ID. */
+    bool success = true;
+
+    for (const auto &member : nhgm_ids)
```

Comment:
The next hop group will not be added to the map of synced groups if this fails. This will cause the unique pointer to be tidied up and the group object freed, and it'll be tidied up over the SAI as part of that